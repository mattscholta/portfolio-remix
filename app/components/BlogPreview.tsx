import { Link } from "@remix-run/react";
import classnames from "classnames";

export interface BlogPreviewProps {
  className?: string;
  content: string;
  date: string;
  featured?: boolean;
  heading?: string;
  slug: string;
  title: string;
  image: string;
}

export const BlogPreview = (props: BlogPreviewProps) => {
  const { className, date, heading = "h3", slug, image, title } = props;

  // Setup
  const Heading = heading as any;
  const created = new Date(date);
  const dateString = created.toLocaleDateString("en-US", {
    weekday: undefined,
    year: "numeric",
    month: "2-digit",
    day: "numeric"
  });

  return (
    <Link
      className={classnames("blog-preview relative flex flex-col", className)}
      prefetch="intent"
      to={`/blog/${slug}`}
    >
      {/* TODO: Add a nice CSS transition here */}
      <img alt={title} height="auto" loading="lazy" src={image} width="auto" />

      <Heading className="mt-4 mb-2 text-2xl">{title}</Heading>
      <div className="font-font-monospace text-sm">{dateString}</div>
      {/* <div dangerouslySetInnerHTML={{ __html: content }} /> */}
    </Link>
  );
};
