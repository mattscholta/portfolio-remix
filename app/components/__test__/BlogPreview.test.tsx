import { expect, test } from "vitest";
import { render } from "@testing-library/react";
import type { RenderResult } from "@testing-library/react";

import { BlogPreview } from "../BlogPreview";
import type { BlogPreviewProps } from "../BlogPreview";
import { MemoryRouter } from "react-router";

describe("BlogPreview Component", () => {
  let props: BlogPreviewProps;
  let component: RenderResult;

  const createWrapper = () => {
    component = render(
      <MemoryRouter>
        <BlogPreview {...props} />
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    props = {
      content: "<p>Hello World</p>",
      date: "2020-01-01",
      image: "https://example.com/image.jpg",
      slug: "hello-world",
      title: "Hello World"
    };

    createWrapper();
  });

  test("should render our BlogPreview", () => {
    expect(component.baseElement).toMatchSnapshot();

    const elDate = component.getByText("01/01/2020"); // props.post.date
    const elHeading = component.getByText(props.title);
    const elImages = component.getAllByAltText(props.title ?? "_no_alt_");

    expect(elDate).toBeInTheDocument();
    expect(elHeading).toBeInTheDocument();
    expect(elImages[0]).toBeInTheDocument();
  });
});
