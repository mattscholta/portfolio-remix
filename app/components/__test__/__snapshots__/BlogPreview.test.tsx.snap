// Vitest Snapshot v1

exports[`BlogPreview Component > should render our BlogPreview 1`] = `
<body>
  <div>
    <a
      class="blog-preview relative flex flex-col"
      href="/blog/hello-world"
    >
      <img
        alt="Hello World"
        height="auto"
        loading="lazy"
        src="https://example.com/image.jpg"
        width="auto"
      />
      <h3
        class="mt-4 mb-2 text-2xl"
      >
        Hello World
      </h3>
      <div
        class="font-font-monospace text-sm"
      >
        01/01/2020
      </div>
    </a>
  </div>
</body>
`;
